src/System.Buffers/src/System/Buffers/DefaultArrayPoolBucket.cs

```diff
@@ -48,9 +54,19 @@ internal T[] Rent()
                 // Check if all of our buffers have been rented
                 if (_index < _data.Length)
                 {
-                    buffer = _data[_index] ?? new T[_bufferLength];
+                    buffer = _data[_index];
+                    if (buffer == null)
+                    {
+                        buffer = new T[_bufferLength];
+                        _eventSource.BufferCreated(buffer.GetHashCode(), _bufferLength, GetHashCode());
+                    }
                     _data[_index++] = null;
                 }
+                else if (_exhaustedEventSent == false)
+                {
+                    _eventSource.BucketExhausted(GetHashCode(), _bufferLength, _data.Length);
+                    _exhaustedEventSent = true;
+                }
```

Comment:
Not necessarily; BufferCreated is only fired when a buffer is **initially** created, since each buffer in a bucket is lazily initialized. Once a BufferBucket is exhausted, firing off the BucketExhausted event, we will try the next Bucket higher; if there isn't a higher sized bucket then we will allocate a new Buffer, firing a BufferAllocated event. The BucketExhausted event is meant to trace if your configuration option for ```MaxBuffersPerBucket``` is too small or not